### ISSUE TITLE
`r2 -d /bin/ls` on grsecurity + PaX system

### ISSUE THREAD

**Issue by XVilka** [radare/radare2#4972]
```
$ r2 -d /bin/ls
Process with PID 31576 started...
attach 31576 31576
r_debug_map_new: error assert(           0>=0)
Assuming filepath /bin/ls
asm.bits 32
r_debug_map_new: error assert(           0>=0)
[0xac1e2aa0]>
```
After that basic stuff, like `ds` or `dso` works good, so it's minor issue

**Comment by jduck** [comment]
The question here is: what is calling r_debug_map_new with 0 for both addr and addr_end...

**Comment by jduck** [comment]
OMG I just looked at ```r_debug_native_map_get``` and wanted to throw up.

**Comment by jduck** [comment]
Can you provide the output from /proc/31576/maps ?

**Comment by XVilka** [comment]
I can provide it only after tomorrow - I don't have this machine near me
right now.

**Comment by jduck** [comment]
Okay, whenever you're ready =)

**Comment by dkasak** [comment]
Since I have the same setup (grsec kernel with PaX), I'm able to provide the requested information:

```
$ r2 -d /bin/ls
Process with PID 2886 started...
attach 2886 2886
r_debug_map_new: error assert(			0>=0)
Assuming filepath /bin/ls
asm.bits 64
r_debug_map_new: error assert(			0>=0)
 -- THE CAKE IS A PIE
```

```
00000000-00000000 r-xp 00000000 fe:01 1058864                            /usr/bin/ls
00000000-00000000 rw-p 00000000 fe:01 1058864                            /usr/bin/ls
00000000-00000000 rw-p 00000000 00:00 0 
00000000-00000000 ---p 00000000 00:00 0 
00000000-00000000 rw-p 00000000 00:00 0                                  [heap]
00000000-00000000 r-xp 00000000 fe:01 1051815                            /usr/lib/ld-2.23.so
00000000-00000000 r--p 00000000 00:00 0                                  [vvar]
00000000-00000000 r-xp 00000000 00:00 0                                  [vdso]
00000000-00000000 rw-p 00000000 fe:01 1051815                            /usr/lib/ld-2.23.so
00000000-00000000 rw-p 00000000 00:00 0 
00000000-00000000 rw-p 00000000 00:00 0                                  [stack]
ffffffffff600000-ffffffffff601000 r--p 00000000 00:00 0                  [vsyscall]
```

**Comment by jduck** [comment]
Yeah, it seems impossible to fix this with code changes. We'd have to scan all of address space for ELF headers and that's error prone (think about faking it inside a heap block or such). I think this will have to be a WONTFIX but should more than likely be addressed in documentation (cc @Maijin)

**Comment by jduck** [comment]
Well, at least I will make it fail more gracefully and with better reasoning than the "error assert(" =)

**Comment by dkasak** [comment]
It seems that breakpoints don't work unless MPROTECT is also turned off, so that workaround above should also contain `m` at least: `setfattr -n user.pax.flags -v mr <executable path>`

**Comment by radare** [comment]
well.. there's a fallback implementation in r2 to solve this when there's no /proc mounted. the code was in r1... and iirc its also in r2 when it checks if /proc exists. it basically walks from 0 to UT32_MAX (on 32bit systems) every page size reading few bytes . if read results in not error it marks that as a page. so it is able to grab the memory pages even with those protections enabled.

**Comment by jduck** [comment]
That's an interesting approach but touching over a million pages seems a little bit sub-optimal.

**Comment by jduck** [comment]
And on 64-bit this seems totally ridiculous too.

**Comment by radare** [comment]
yep it's sub-optimal, but it's the first solution that came to my mind. and the result can be cached. any other way to bypass this protection? anyway, is /proc/pid/mem also banned here?

**Comment by radare** [comment]
what happend with gdb or lldb?

**Comment by radare** [comment]
can be disabled for just one process . See this:

https://wiki.gentoo.org/wiki/Hardened/Debugging <https://wiki.gentoo.org/wiki/Hardened/Debugging>


thanks @revskills
